### PR TITLE
Fix

### PR DESCRIPTION
**Pull request by alenworld**
Removed route /v1/user Create.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,33 +1,38 @@
...
   },
-  "homepage": "https://github.com/ChechaValerii/intership-template#readme",
+  "homepage": "https://github.com/alenworld/intership-template#readme",
   "dependencies": {
-    "@hapi/joi": "^17.1.1",
+    "bcrypt": "^5.0.1",
     "compression": "1.7.4",
     "cookie-parser": "1.4.6",
     "cors": "2.8.5",
+    "dotenv": "^10.0.0",
```
Comment (v-checha) [inline package.json]: Please remove corners

File: package.json [hunk 2]
```diff
@@ -1,33 +1,38 @@
...
     "express": "4.17.1",
     "helmet": "4.6.0",
-    "mongoose": "6.0.14"
+    "joi": "^17.1.1",
```
Comment (v-checha) [inline package.json]: the same

File: package.json [hunk 3]
```diff
@@ -1,33 +1,38 @@
...
+    "jsonwebtoken": "^8.5.1",
+    "mongoose": "6.0.14",
+    "uuid": "^8.3.2"
   },
   "devDependencies": {
     "eslint": "^8.3.0",
     "eslint-config-airbnb-base": "^15.0.0",
-    "eslint-plugin-import": "^2.25.3"
+    "eslint-plugin-import": "^2.25.3",
```
Comment (v-checha) [inline package.json]: the same

File: src/components/Auth/helper.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  };
+
+  const options = { expiresIn: tokens.access.expiresIn };
+
+  return jwt.sign(payload, secret, options);
+};
+
+const generateRefreshToken = () => {
+  const payload = {
+    id: uuid(),
```
Comment (v-checha) [inline src/components/Auth/helper.js:19]: better if access token and refresh will have the same payload

File: src/components/Auth/helper.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+};
+
+const replaceDbRefreshToken = async (tokenId, userId) => {
+  try {
+    const oldRefreshToken = await AuthService.removeRefreshToken(userId);
+    if (oldRefreshToken) {
+      throw new Error('Invalid token!');
+    }
+    const updatedRefreshToken = await AuthService.createRefreshToken(tokenId, userId);
+    return updatedRefreshToken;
```
Comment (v-checha) [inline src/components/Auth/helper.js]: must be `return AuthService.createRefreshToken(tokenId, userId)`

File: src/components/Auth/helper.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  } catch (error) {
```
Comment (v-checha) [inline src/components/Auth/helper.js]: remove try catch after replace to return promise style and catch errors in controller

File: src/components/Auth/index.js [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+const jwt = require('jsonwebtoken');
+const AuthService = require('./service');
+const AuthHelper = require('./helper');
+const UserValidation = require('../User/validation');
+const UserService = require('../User/service');
+const ValidationError = require('../../error/ValidationError');
+const { secret } = require('../../config/env').JWT;
+
+const updateTokens = async (userId) => {
+  const accessToken = await AuthHelper.generateAccessToken(userId);
```
Comment (v-checha) [inline src/components/Auth/index.js]: generateAccessToken is sync code, no need await 
the same in generateRefreshToken and replaceDbRefreshToken

File: src/components/Auth/index.js [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+  const refreshToken = await AuthHelper.generateRefreshToken();
+  const newestTokens = await AuthHelper.replaceDbRefreshToken(refreshToken.id, userId)
+    .then(() => ({
```
Comment (v-checha) [inline src/components/Auth/index.js]: no need then, replaceDbRefreshToken is not a promise

File: src/components/Auth/index.js [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+/**
+ * @function
+ * @param {express.Request} req
+ * @param {express.Response} res
+ * @param {express.NextFunction} next
+ * @returns {Promise < void >}
+ */
+const signUp = async (req, res, next) => {
+  try {
+    const { error } = UserValidation.create(req.body);
```
Comment (v-checha) [inline src/components/Auth/index.js]: get value after validation data `const { error, value  } = UserValidation.create(req.body);`

File: src/components/Auth/index.js [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+
+    if (error) {
+      throw new ValidationError(error.details);
+    }
+
+    const isUser = await AuthService.findByEmail(req.body.email);
```
Comment (v-checha) [inline src/components/Auth/index.js]: use `value`  instead of req.body.email

File: src/components/Auth/index.js [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+    if (isUser === null) {
+      const userData = {
+        firstName: req.body.firstName,
```
Comment (v-checha) [inline src/components/Auth/index.js]: use value

File: src/components/Auth/index.js [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+        lastName: req.body.lastName,
+        email: req.body.email,
+        password: bcrypt.hashSync(req.body.password, 10),
+      };
+
+      const user = await UserService.create(userData);
+      res.status(200).json({ data: user });
+    } else {
+      res.status(400).json({ message: 'Email already exists!' });
```
Comment (v-checha) [inline src/components/Auth/index.js]: rename message to error

File: src/components/Auth/index.js [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+    }
+  } catch (error) {
+    res.status(500).json({
```
Comment (v-checha) [inline src/components/Auth/index.js:52]: did not handle validation error like in example

File: src/components/Auth/index.js [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+/**
+ * @function
+ * @param {express.Request} req
+ * @param {express.Response} res
+ * @param {express.NextFunction} next
+ * @returns {Promise < void >}
+ */
+const signIn = async (req, res, next) => {
+  try {
+    const { email, password } = req.body;
```
Comment (v-checha) [inline src/components/Auth/index.js]: please validate params with Joi

File: src/components/Auth/index.js [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+    } else {
+      const isValid = bcrypt.compareSync(password, user.password);
+      if (isValid) {
+        updateTokens(user._id).then((tokens) => res.json(tokens));
+      } else {
+        res.status(401).json({ message: 'Invalid credentials!' });
+      }
+    }
+  } catch (error) {
+    res.status(500).json({
```
Comment (v-checha) [inline src/components/Auth/index.js:96]: catch validation error like in example

File: src/components/Auth/index.js [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+      message: error.name,
+      details: error.message,
+    });
+    next(error);
+  }
+};
+
+const refreshTokens = async (req, res) => {
+  const { refreshToken } = req.body;
```
Comment (v-checha) [inline src/components/Auth/index.js]: please validate params with Joi

File: src/components/Auth/index.js [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+
+  let payload;
+
+  try {
+    payload = jwt.verify(refreshToken, secret);
+    if (payload.type !== 'refresh') {
+      res.status(400).json({ message: 'Invalid token!' });
```
Comment (v-checha) [inline src/components/Auth/index.js:112]: `return res.status(400).json({ message: 'Invalid token!' });`

File: src/components/Auth/index.js [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+    }
+  } catch (error) {
+    if (error instanceof jwt.TokenExpiredError) {
+      res.status(400).json({ message: 'Token expired!' });
+    } else if (error instanceof jwt.JsonWebTokenError) {
+      res.status(400).json({ message: 'Invalid token!' });
+    }
+  }
+
+  const findToken = await AuthService.findTokenById({ tokenId: payload.id });
```
Comment (v-checha) [inline src/components/Auth/index.js]: `const refreshToken = `

File: src/components/Auth/service.js [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+
+/**
+ * @exports
+ * @method findByEmail
+ * @param {string} email
+ * @summary get a user
+ * @returns {Promise<UserModel>}
+ */
+function findByEmail(email) {
+  return UserModel.findOne({ email }).exec();
```
Comment (v-checha) [inline src/components/Auth/service.js]: `return UserModel.findOne({ email }).lean().exec();`
use .lean().exec(); in all methods with database or only lean if you no need mongoose instance

File: src/config/connection.js [hunk 20]
```diff
@@ -1,12 +1,23 @@
...
+   */
 
-const connectOptions = {
+  const options = {
     useNewUrlParser: true,
     useUnifiedTopology: true,
+  };
+  mongoose.connect(MONGO_URI, options);
+
+  require('../components/User/model');
```
Comment (v-checha) [inline src/config/connection.js]: not clear why you need this?

File: package.json (hunk 1 above)
Comment (alenworld) [reply package.json]: Done

File: package.json (hunk 2 above)
Comment (alenworld) [reply package.json]: Done

File: package.json (hunk 3 above)
Comment (alenworld) [reply package.json]: Done

File: src/components/Auth/helper.js (hunk 5 above)
Comment (alenworld) [reply src/components/Auth/helper.js]: Done

File: src/components/Auth/helper.js (hunk 6 above)
Comment (alenworld) [reply src/components/Auth/helper.js]: Done

File: src/components/Auth/index.js (hunk 7 above)
Comment (alenworld) [reply src/components/Auth/index.js]: Done

File: src/components/Auth/index.js (hunk 8 above)
Comment (alenworld) [reply src/components/Auth/index.js]: Done

File: src/components/Auth/index.js (hunk 9 above)
Comment (alenworld) [reply src/components/Auth/index.js]: Done

File: src/components/Auth/index.js (hunk 10 above)
Comment (alenworld) [reply src/components/Auth/index.js]: Done

File: src/components/Auth/index.js (hunk 11 above)
Comment (alenworld) [reply src/components/Auth/index.js]: Done

File: src/components/Auth/service.js (hunk 19 above)
Comment (alenworld) [reply src/components/Auth/service.js]: Done

File: src/components/Auth/index.js (hunk 18 above)
Comment (alenworld) [reply src/components/Auth/index.js]: Done

**Pull request closed without merging**